### PR TITLE
init: Add progress indicators during long running updates

### PR DESCRIPTION
**Pull request by MilhouseVH**
Merge this AFTER #27!

This adds various progress indicators during long running update steps:

* a simple spinner `|`, `/`, `-`, etc. for any indeterminately long process (ie. gunzip as we don't know the uncompressed file size)
* a percentage complete indicator during file copies when the file size is known `Updating KERNEL... 28%`
* a countdown indicator, `rebooting in 10 seconds... 10`, `rebooting in 10 seconds... 9` ... `rebooting in 10 seconds... NOW`.

Also improve quality of some error messages, taking the opportunity to explain what options the user has when failing "Check size", and correcting some code indentation issues.

### CODE REVIEW

**Comment by MilhouseVH** [comment]
Rebased after #27.

**Comment by lrusak** [comment]
Looks good, are you happy with this or does it need more testing?

**Comment by MilhouseVH** [comment]
I'm in the process of uploading builds for internal testing, if we're all happy then we can merge this later this evening.

**Comment by stefansaraev** [comment]
here an idea for improvement: http://sprunge.us/cPEZ 

functions can be moved to separate file, which can be re-used in init and scripts like fs-resize.

**Comment by MilhouseVH** [comment]
Thanks, I'd really like to split some of the functions in `init` out to separate files, if there's no objections I can make that change.

For `init` I'm not sure we want to execute the commands (`dd` etc.) within the `start_progress()` call, as in some cases (ie. `md5sum`, `gunzip`) the exit value of the command determines the value displayed by `stop_progress()`, ie. "OK" or "FAILED" etc.  It's possible we could pass the command in some situations, but not easily in others.

For `fs-resize` I can see it being used there, so I'll see about making the change to this PR.

**Comment by stefansaraev** [comment]
```
eval $cmd
local ret=$?
stop_progress "something"
return $ret
```

then you can check [ "$?" -eq something ] after a progress task is stopped/done

**Comment by MilhouseVH** [comment]
@stefansaraev : what do you think of this? https://gist.github.com/MilhouseVH/67dcfc17175f476300cc

Test script: https://gist.github.com/MilhouseVH/2ca3459e29e6e9fa6033

I think that covers all the requirements...

Later today I'll update `init` and `fs-resize`.

**Comment by stefansaraev** [comment]
looks good.

**Comment by MilhouseVH** [comment]
`fs-resize`: https://gist.github.com/MilhouseVH/c73a069849a7712adbe0

Feel free to suggest better descriptions for the resize steps - they're the best I could come up with based on my interpretation of what the `parted`/`e2fsck`/`resizefs` commands are doing.

Back later.

**Comment by vpeter4** [comment]
Why exactly resize2fs needs to be called anyway? Partition is already deleted and created new one.
If really needed then e2fsck should be called after also?

**Comment by stefansaraev** [comment]
> Partition is already deleted and created new one

partition != filesystem

> If really needed then e2fsck should be called after also

no. before is enough.

File: packages/sysutils/busybox/scripts/init [hunk 1]
```diff
@@ -191,6 +203,112 @@
...
+  }
+
+  # Use this task when transferring a file of known size
+  ProgressTask_Percent() {
+    local msg="$1" filename="$2" fsize=$3 csize donemsg
+
+    echo -n "${msg}    "
+
+    while [ ! -f /tmp/progress.stop ]; do
+      [ -f ${filename} ] && csize=$(stat -t "${filename}" | awk '{print $2}') || csize=0
```
Comment (stefansaraev) [inline packages/sysutils/busybox/scripts/init]: avoiding awk usage would be nice.
```
    [ -f ${filename} ] && csize=$(stat -t "${filename}" | cut -d" " -f2) || csize=0
    printf "\b\b\b\b%3d%%" $(($csize * 100 / $fsize))
```

**Comment by vpeter4** [comment]
And if resize2fs fail somehow you end with broken filesystem.

File: packages/sysutils/busybox/scripts/init [hunk 2]
```diff
@@ -191,6 +203,112 @@
...
+
+  # Use this task to countdown a number of seconds
+  # (needs more work for durations > 99 seconds)
+  ProgressTask_Countdown() {
+    local msg="$1" countfrom=$2 donemsg="$3"
+
+    echo -n "${msg}  "
+
+    while [ ${countfrom} -gt 0 ]; do
+      echo ${countfrom} | awk '{ printf "\b\b%2d", $1 }'
```
Comment (stefansaraev) [inline packages/sysutils/busybox/scripts/init]: same here, could be
```
    printf "\b\b%2d" ${countfrom}
```

**Comment by stefansaraev** [comment]
same for if resizefs is not called, you end with with broken system. why are we even discussing this here in this pr?

**Comment by MilhouseVH** [comment]
I've pushed some updates.

Yep, those echo/awk commands will have to stay as they are unless we want to enable more busybox applets, which probably isn't worth it.

File: packages/sysutils/busybox/scripts/functions [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+  # This file is used to indicate a metter is running/active
+  echo > /tmp/progress.run
+
+  case "$1" in
+    spinner)
+      ProgressTask_Spinner "$2" &
+      if [ -n "$3" ]; then
+        eval "$3"
+        cmdresult=$?
+        [ $cmdresult -eq 0 ] && StopProgress "${4}" || StopProgress "${5:-$4}"
```
Comment (stefansaraev) [inline packages/sysutils/busybox/scripts/functions]: if command is optional, you'd never cancel the spinner progress here. same for percent below.

Comment (MilhouseVH) [inline packages/sysutils/busybox/scripts/functions]: And nor would you want to - take a look at https://github.com/MilhouseVH/LibreELEC.tv/blob/init_progress/packages/sysutils/busybox/scripts/init#L615-L620.

Comment (MilhouseVH) [inline packages/sysutils/busybox/scripts/functions]: The way the progress meter is now written you can have it do everything for you if the command is relatively simple, or for more complex commands (or coding clarity) you can control the start and stop separately.

**Comment by stefansaraev** [comment]
btw, for fs-resize, /tmp may not be available. better use /dev

**Comment by MilhouseVH** [comment]
Really, no /tmp during fs-resize? Bummer. I'll do some testing and figure out a solution.

**Comment by MilhouseVH** [comment]
Updated to use `/dev/.progress` in place of `/tmp/`

Tested `fs-resize` (disk image) and upgrading an existing installation (from `.img.gz`). Both worked fine on an RPi3.

**Comment by stefansaraev** [comment]
//bikeshed mode on
you could just use /dev/.progress.{run,stop} 
//bikeshed mode off

thanks for this, users will be happy as they know their box is not dead.

https://github.com/LibreELEC/LibreELEC.tv/blob/master/packages/mediacenter/LibreELEC-settings/scripts/backup-restore this needs some love, too.

File: packages/sysutils/busybox/scripts/init [hunk 4]
```diff
@@ -24,6 +24,7 @@
   /bin/busybox mkdir -p /dev
   /bin/busybox mkdir -p /proc
   /bin/busybox mkdir -p /sys
+  /bin/busybox mkdir -p /tmp
```
Comment (stefansaraev) [inline packages/sysutils/busybox/scripts/init]: not needed anymore

Comment (MilhouseVH) [inline packages/sysutils/busybox/scripts/init]: It is - https://github.com/LibreELEC/LibreELEC.tv/pull/44/files#diff-708fce2b98cae7109f84bb7011142f00R635

I could have changed it, but I thought having /tmp around might prove useful in future...

Comment (stefansaraev) [inline packages/sysutils/busybox/scripts/init]: right. missed. sorry for the noise.

**Comment by MilhouseVH** [comment]
> https://github.com/LibreELEC/LibreELEC.tv/blob/master/packages/mediacenter/LibreELEC-settings/scripts/backup-restore this needs some love, too.

Good idea, I think I'll handle that as a separate PR.

**Pull request merged**